### ISSUE TITLE
Question: why does prepare_model_for_kbit_training cast input **and** output embeddings to float32?

### ISSUE THREAD

**Issue by Birch-san** [huggingface/peft#816]
### System Info

System information not relevant. I am asking about an implementation decision.

### Who can help?

@pacman100 @younesbelkada  

I'd appreciate an explanation of the casting performed by [`prepare_model_for_kbit_training`](https://github.com/huggingface/peft/blob/7d44026dea18221e33affaa4868c908ef2c33f54/src/peft/utils/other.py#L79).
_recent activity: https://github.com/huggingface/peft/pull/389._

reading the docstring: I understand that it _intends_ to cast layernorms and lm_head to float32.

in practice however ([qlora.py](https://github.com/artidoro/qlora/blob/7f4e95a68dc076bea9b3a413d2b512eca6d004e5/qlora.py#L377)): I observe that it casts _more layers than that._

it casts the following to float32:

- layernorms
- `model.get_input_embeddings().weight` (`embed_tokens: Embedding`)
- `model.get_output_embeddings().weight` (`lm_head: Linear`)

was it _intentional_ to upcast **both** the input **and** output embeddings?

in a model with _tied_ embeddings (such as MPT): `model.get_input_embeddings()` and `model.get_output_embeddings()` share the same weights, `embed_tokens: Embedding`. so when you say you're upcasting lm_head, I wonder whether that means you're actually upcasting `embed_tokens`.

in a model with _untied_ embeddings (such as Llama, Falcon, Pythia): `lm_head: Linear` and `embed_tokens: Embedding` are separate weights and not even the same type of torch module.

so, for Llama-type models: we are casting both `lm_head: Linear` **and** `embed_tokens: Embedding` to float32. perhaps that's unnecessary? could it be that actually _only one of these needs to be float32?_

I'm asking because _both_ of these are **big** layers. (on Pythia, either of these would be 788MB in float32).

if it turns out that _only the `Embedding`_ or _only the `Linear`_ needs to be float32, then a lot of memory could be saved.

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

https://github.com/artidoro/qlora

```bash
python qlora.py huggyllama/llama-7b
```

Relevant line is here:  
https://github.com/artidoro/qlora/blob/7f4e95a68dc076bea9b3a413d2b512eca6d004e5/qlora.py#L377

### Expected behavior

`prepare_model_for_kbit_training`'s docstring should explain whether it intends to upcast input embeddings **and** output embeddings, or just one of those.

if only one of those upcasts is intended: `prepare_model_for_kbit_training` needs changing, to be more selective in which layers it upcasts.

**Comment by BenjaminBossan** [comment]
I agree that there is a mismatch between what happens in the function body and what the docstring says. When looking at #389, it does seem intentional though. Maybe it would be possible to make that change optional, not sure. Let's wait what the others have to say, as I don't know the full context.

**Comment by BenjaminBossan** [comment]
I wonder if we can re-use, or take inspiration from [this accelerate function](https://github.com/huggingface/accelerate/blob/058a3546ea588abf57de3705409c250cc52423e7/src/accelerate/utils/bnb.py#L372).

**Comment by PanQiWei** [comment]
My point of view is up-cast to fp32 is for the purpose of stablelize training loss, this may slightly increase VRAM usage but not slowdown training speed. Not sure if I understand correct, also looking forward an explaination.

**Comment by ecolss** [comment]
could someone give us a proper explanation?

**Comment by andeyeluguo** [comment]
somebody said we can just remove the call
[https://github.com/huggingface/peft/issues/1149](url)

**Comment by beingdutta** [comment]
That method internally casts the LayerNorm, embeddings and other sensitive layers to fp16/bf16, for making stuff numerically stable. If you use this method is mostly used when you have quantised a model. Now, if you quantise a model, all the layers are downcasted to 4-bit or 8-bit, and that is highly unstable for training, so it intentionally upcasts at least those layers mentioned to maintain numerical stability.
